client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-prefetch-images.ts

```diff
@@ -0,0 +1,47 @@
+import { useEffect, useState } from 'react';
+
+const IMAGE_URLS = [
+	'https://dotcompatterns.files.wordpress.com/2023/03/header-mountains.jpeg',
+
+	'https://blankcanvas3demo.files.wordpress.com/2023/03/mountains_landscape.jpeg?w=1024',
+	'https://blankcanvas3demo.files.wordpress.com/2023/03/mountains_landscape.jpeg',
+	'https://dotcompatterns.files.wordpress.com/2023/03/mountains_landscape.jpeg?w=1024',
+	'https://dotcompatterns.files.wordpress.com/2023/03/mountains_landscape.jpeg',
+
+	'https://dotcompatterns.files.wordpress.com/2023/03/mountains_portrait.jpeg?w=640',
+	'https://dotcompatterns.files.wordpress.com/2023/03/mountains_portrait.jpeg',
+
+	'https://dotcompatterns.files.wordpress.com/2023/03/mountains_square.jpeg?w=800',
+];
+
+/**
+ * Prefetches the images that can be bottleneck for the pattern rendering.
+ */
+export const usePrefetchImages = () => {
+	const [ links, setLinks ] = useState< HTMLLinkElement[] >( [] );
+
+	const prefetchImage = ( url: string ) => {
+		const link = document.createElement( 'link' );
+		link.rel = 'prefetch';
+		link.as = 'image';
+		link.href = url;
+		document.head.appendChild( link );
+		setLinks( ( links ) => [ ...links, link ] );
+	};
+
+	const removeLinks = ( link: HTMLLinkElement ) => {
+		document.head.removeChild( link );
+	};
+
+	useEffect( () => {
+		IMAGE_URLS.forEach( ( url ) => {
+			prefetchImage( url );
+		} );
+	}, [] );
+
+	useEffect( () => {
+		return () => {
+			links.forEach( ( link ) => removeLinks( link ) );
```

Comment:
This might not be intuitive as we try to remove previous links whenever the links change. Now, we call `setLinks` in the loop so react will batch updating the links at once. However, if there is any delay accidentally, some of the prefetch links will be removed immediately.

Reply:
> we try to remove previous links whenever the links change

Ahh, thank you for pointing that out.
In this case, I might better use `useRef` or something.

Reply:
I guess leaving the deps as `[]` is enough since we want to remove the links only when umount, right?

Reply:
>I guess leaving the deps as [] is enough 

I confirmed this does not respect updated `links` because it just uses the value available at the initial render. (`links` in `links.forEach( ( link ) => removeLinks( link ) );` is stay empty `[]`.)